src/main/kotlin/io/cloudflight/gradle/autoconfigure/kotlin/KotlinConfigurePlugin.kt

```diff
@@ -0,0 +1,88 @@
+package io.cloudflight.gradle.autoconfigure.kotlin
+
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.apply
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.create
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.getByType
+import io.cloudflight.gradle.autoconfigure.java.JavaConfigurePlugin
+import io.cloudflight.gradle.autoconfigure.java.JavaConfigurePluginExtension
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.jvm.toolchain.JavaToolchainSpec
+import org.jetbrains.kotlin.allopen.gradle.AllOpenExtension
+import org.jetbrains.kotlin.allopen.gradle.SpringGradleSubplugin
+import org.jetbrains.kotlin.gradle.dsl.KotlinProjectExtension
+import org.jetbrains.kotlin.gradle.internal.Kapt3GradleSubplugin
+import org.jetbrains.kotlin.gradle.plugin.KotlinPluginWrapper
+import org.jetbrains.kotlin.gradle.plugin.getKotlinPluginVersion
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.noarg.gradle.KotlinJpaSubplugin
+import org.slf4j.LoggerFactory
+
+class KotlinConfigurePlugin : Plugin<Project> {
+    override fun apply(project: Project) {
+        project.plugins.apply(JavaConfigurePlugin::class)
+
+        project.plugins.apply(KotlinPluginWrapper::class)
+
+        project.plugins.apply(Kapt3GradleSubplugin::class)
+        project.plugins.apply(KotlinJpaSubplugin::class) // TODO only if JPA entities detected on classpath
+        project.plugins.apply(SpringGradleSubplugin::class)   // TODO only when there is Spring on the classpath
+
+        val extensions = project.extensions
+        val tasks = project.tasks
+
+        extensions.create(EXTENSION_NAME, KotlinConfigurePluginExtension::class).apply {
+            kotlinVersion.convention(project.getKotlinPluginVersion())
+        }
+
+        extensions.getByType(AllOpenExtension::class.java).apply {
+            annotation("javax.persistence.Entity")
+            annotation("javax.persistence.MappedSuperclass")
+        }
+
+        val kotlin = extensions.getByType(KotlinProjectExtension::class.java)
+        kotlin.sourceSets.maybeCreate("main").dependencies {
+            // we need to do that lazily here in order to evaluate the KotlinConfigurePluginExtension
+            // not too early to fetch the configured kotlin version
+            this.implementation(project.provider {
```

Comment:
instead of using an explicit provider you should use a [mapped property](https://docs.gradle.org/current/userguide/lazy_configuration.html#creating_property_provider).   
`A Provider can also be created by the factory method [ProviderFactory.provider(Callable)](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ProviderFactory.html#provider-java.util.concurrent.Callable-). You should prefer using map() instead, as this has some useful benefits, which we will see later.`

```kotlin
kotlinConfigureExtension.kotlinVersion.map { "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$it" }
```